status-im/status-react
# fix #1050 and introduce money ns that uses bignumber for arithmetic

fixes #1050 

Introduce a money namespace that solves a bunch of issues related to significant numbers being preserved and converted correctly.

### Summary:
The BigNumber version included in web3 sometimes hangs when dividing large numbers. Hence we want to use these functions instead of fromWei etc, which come bundled with web3. See
https://github.com/MikeMcl/bignumber.js/issues/120 for this regression being introduced in some JS environments. It is fixed in the MikeMcl/bignumber.js repo, but not in the web3 BigNumber fork:
https://github.com/ethereum/web3.js/issues/877

Additionally, while it is possible to use the BigNumber constructor without stringifying the number, this only works up to some 15 significant digits:
https://github.com/MikeMcl/bignumber.js/issues/120

Lastly, notice the bad rounding for native Javascript numbers above 17 digits that may result in errors earlier up the call chain. Ideally all money-related sensitive functions should be moved into this namespace to check for such matters:
(str 111122223333441239) => "111122223333441230"

### Steps to test:
- Open Status (iOS)
- Ensure you Ether on your account
- /send PERSON 0.111122223333444
- Notice that app no longer crashes

[comment]: # (PRs will only be accepted if squashed into single commit.)
status: ready

Tested on iOS simulator. As this change involves change of packages etc, and Web3 already includes their version of BigNumber, we have to be careful to test it on real devices to make sure the wrong version of the library isn't being called.
